File: ios/App/App/Features/RichChat/TerminalCursorFrameDecoder.swift
import Foundation

/// Hand-written coordinator: decodes a `terminal-output` cursor frame or a
/// ready `terminal-watch-result` baseline into the generated
/// `TerminalCursorFrame`. The reconciliation decision itself is the generated
/// terminal-cursor machine (`TerminalCursorMachine.swift`); only this RichJSON
/// plumbing stays app-owned (the TS reference is
/// `terminalCursorMachine.ts#decodeTerminalCursorFrameMessage`).
enum TerminalCursorFrameDecoder {
  static func decode(_ value: RichJSON) throws -> TerminalCursorFrame {
    guard let object = value.objectValue,
      let type = RichDecoding.requiredString(object, "type"),
      let terminalID = RichDecoding.requiredString(object, "id", allowEmpty: false)
    else { throw RichDomainDecodeError.invalidTerminalFrame }
    if type == "terminal-output" {
      guard let data = RichDecoding.requiredString(object, "data"),
        let sync = object["cursorSync"]?.objectValue,
        let generation = RichDecoding.requiredString(sync, "generation", allowEmpty: false)
      else { throw RichDomainDecodeError.invalidTerminalFrame }
      return try decodeRange(
        kind: .output,
        terminalID: terminalID,
        generation: generation,
        data: data,
        sync: sync
      )
    }
    if type == "terminal-watch-result" {
      guard let sync = object["cursorSync"]?.objectValue,
        sync["version"]?.exactInt64Value == 1,
        let watchID = RichDecoding.requiredString(sync, "watchId", allowEmpty: false),
        let result = sync["result"]?.objectValue,
        RichDecoding.requiredString(result, "status") == "ready",
        let data = RichDecoding.requiredString(result, "data"),
        let from = result["fromCursor"]?.exactInt64Value,
        let to = result["toCursor"]?.exactInt64Value,
        result.keys.contains("generation")
      else { throw RichDomainDecodeError.invalidTerminalFrame }
      let generation: String?
      if result["generation"] == .null {
        generation = nil
      } else if let value = result["generation"]?.stringValue, !value.isEmpty {
        generation = value
      } else {
        throw RichDomainDecodeError.invalidTerminalFrame
      }
      let frame = TerminalCursorFrame(
        kind: .baseline,
        terminalID: terminalID,
        watchID: watchID,
        generation: generation,
        fromCursor: from,
        toCursor: to,
        data: data
      )
      guard TerminalCursorReconciler.isValid(frame) else {
        throw RichDomainDecodeError.invalidTerminalFrame
      }
      return frame
    }
    throw RichDomainDecodeError.invalidTerminalFrame
  }

  private static func decodeRange(
    kind: TerminalCursorFrameKind,
    terminalID: String,
    generation: String,
    data: String,
    sync: [String: RichJSON]
  ) throws -> TerminalCursorFrame {
    guard sync["version"]?.exactInt64Value == 1,
      let watchID = RichDecoding.requiredString(sync, "watchId", allowEmpty: false),
      let from = sync["fromCursor"]?.exactInt64Value,
      let to = sync["toCursor"]?.exactInt64Value
    else { throw RichDomainDecodeError.invalidTerminalFrame }
    let frame = TerminalCursorFrame(
      kind: kind,
      terminalID: terminalID,
      watchID: watchID,
      generation: generation,
      fromCursor: from,
      toCursor: to,
      data: data
    )
    guard TerminalCursorReconciler.isValid(frame) else {
      throw RichDomainDecodeError.invalidTerminalFrame
    }
    return frame
  }
}

File: android/app/src/main/kotlin/com/poracode/app/chat/TerminalCursorFrameDecoder.kt
package com.poracode.app.chat

import com.poracode.remote.v3.generated.TerminalCursorFrame
import com.poracode.remote.v3.generated.TerminalCursorFrameKind
import com.poracode.remote.v3.generated.TerminalCursorReconciler
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull

/**
 * Hand-written coordinator: decodes a `terminal-output` cursor frame or a ready
 * `terminal-watch-result` baseline into the generated [TerminalCursorFrame].
 * The reconciliation decision itself is the generated terminal-cursor machine;
 * only this JSON plumbing stays app-owned (the TS reference is
 * `terminalCursorMachine.ts#decodeTerminalCursorFrameMessage`).
 */
object TerminalCursorFrameDecoder {
    fun decode(value: JsonElement): TerminalCursorFrame? {
        val objectValue = value.objectOrNull() ?: return null
        val terminalId = objectValue.requiredString("id", allowEmpty = false) ?: return null
        return when (objectValue.requiredString("type")) {
            "terminal-output" -> {
                val data = objectValue.requiredString("data") ?: return null
                val sync = objectValue["cursorSync"]?.objectOrNull() ?: return null
                val generation = sync.requiredString("generation", allowEmpty = false) ?: return null
                decodeRange(TerminalCursorFrameKind.OUTPUT, terminalId, generation, data, sync)
            }
            "terminal-watch-result" -> {
                val sync = objectValue["cursorSync"]?.objectOrNull() ?: return null
                if (sync["version"]?.longOrStrictNull() != 1L) return null
                val watchId = sync.requiredString("watchId", allowEmpty = false) ?: return null
                val result = sync["result"]?.objectOrNull() ?: return null
                if (result.requiredString("status") != "ready") return null
                val generation = when (val raw = result["generation"]) {
                    JsonNull -> null
                    else -> raw?.stringOrNull() ?: return null
                }
                val data = result.requiredString("data") ?: return null
                val from = result["fromCursor"]?.longOrStrictNull() ?: return null
                val to = result["toCursor"]?.longOrStrictNull() ?: return null
                TerminalCursorFrame(
                    TerminalCursorFrameKind.BASELINE,
                    terminalId,
                    watchId,
                    generation,
                    from,
                    to,
                    data,
                ).takeIf(TerminalCursorReconciler::isValid)
            }
            else -> null
        }
    }

    private fun decodeRange(
        kind: TerminalCursorFrameKind,
        terminalId: String,
        generation: String,
        data: String,
        sync: kotlinx.serialization.json.JsonObject,
    ): TerminalCursorFrame? {
        if (sync["version"]?.longOrStrictNull() != 1L) return null
        val watchId = sync.requiredString("watchId", allowEmpty = false) ?: return null
        val from = sync["fromCursor"]?.longOrStrictNull() ?: return null
        val to = sync["toCursor"]?.longOrStrictNull() ?: return null
        return TerminalCursorFrame(kind, terminalId, watchId, generation, from, to, data)
            .takeIf(TerminalCursorReconciler::isValid)
    }
}
